README
xml2po 0.20.10 — © 2004, Danilo Segan

Requirements
............

 * Python
 * libxml2 and libxml2-python
 * xml2po.py (this program)
 * GNU gettext (msgfmt, msgmerge tools)

Quick install
.............

To install in the same prefix as intltool:

 $ ./configure && make && su -c "make install"


About xml2po
............

xml2po is a simple Python program which extracts translatable content
from free-form XML documents and outputs gettext compatible POT
files.

It can work it's magic with most "simple" tags, and for complicated
tags one has to provide a list of all tags which are "final" (that
will be put into one "message" in PO file), "ignored" (skipped over)
and "space preserving".

I will try to provide sane defaults for DocBook documents and other
common document types (like Gnome Summaries, XHTML).  For other kinds
of documents, it is possible to use "-a" (--automatic-tags) to choose
suitable translatable pieces.  It usually works very well.

Preparing a POTemplate from XML file
........................................

 ./xml2po.py -o template.pot file.xml

By default, xml2po treats documents as DocBook.  If that's not
correct, you may choose another "document mode" by passing option "-m
MODE" on the command line.  If there's currently no special module
developed for document type you're trying to translate, use option
"-a" for automatic detection of tags, which will work well even for
simple DocBook documents.

Get back a translation from a translator
........................................

Save a translation to xx.po, where xx is language code of a
translation (i.e. "sr" for Serbian, "de" for German, ...).

Execute:
 ./xml2po.py -p xx.po -o file-xx.xml file.xml

Now you've got a nice and translated XML file in 'file-xx.xml'.

Updating a translation
......................

When base XML file changes, the real advantages of PO files come to
surface.  To merge the translation you need to first produce a new
POT file:
 ./xml2po.py -o newfile.pot newfile.xml

And then you need to use gettext "msgmerge" program to merge the
translation with the new POT file:
 msgmerge -o new-xx.po xx.po newfile.pot && mv new-xx.po xx.po

Alternatively, xml2po provides option "-u" which does exactly these
two steps for you:
 ./xml2po.py -u xx.po newfile.xml

(Note that there has been a change in suggested behaviour between
xml2po versions 1.0.9 and 1.0.10 — previously, you used only "xx",
without ".po", what emulated intltool-update behaviour; users [Ismael
Olea actually :] found that confusing, so I'm suggesting a full
filename now.)

Special handling of document types
..................................

If you want to handle some document types in a special way, you may
need to write short Python module to do it.  Currently, two special
document types come with xml2po: "docbook" and "empty".

Look at files modes/docbook.py and modes/empty.py to see how can one
write new document handling modules.  DocBook module sets "lang"
attribute on "article" element, and also adds translators to
<articleinfo> element in form of <copyright>s.

The basic features every module must provide are following functions:

- getIgnoredTags: list of tags to be skipped over, unless they're
  also "final"

- getFinalTags: list of tags that will end up as separate PO messages

- getPreserveSpaceTags: tags for which spaces should not be
  compressed 

- getStringForTranslators: string that will be added to PO file which
  translators can translate with their names

- getCommentForTranslators: comment that should document what format 
  should translators use to credit themselves

- postProcessXmlTranslation(doc, lang, credits): function which
  receives current document tree (doc), current language (lang) and
  list of translators (credits); if applicable, it should set
  language and credit translators appropriately

If a element name is in both ignored tags and final tags, then it is
treated in a somewhat special way.  It "resets" the state of parser,
but is not included as a separate translable message in PO files.

This is useful with cases like nested lists in other final tags.  In
DocBook, if we have <itemizedlist> inside <para>, we'd preferably
have it replaced in its entirety, instead of getting in message
something like:
 <itemizedlist><listitem><placeholder-1/></listitem></itemizedlist>
(where <placeholder-1/> is a another nested <para>, which is itself a
final tag).  xml2po will try to do best even in cases like this, but
giving it a hand can't hurt.

If you don't want to worry about nesting of tags, option "-a" tries to
detect what nodes are most suitable as final-tags.  This won't work
well if one can nest tags indefinitely (like with DocBook documents,
where one may have "<para>Blah... <orderedlist><listitem>...</para>"
which may become very long string, what defeats the advantages of 
PO files).

Caveats
.......

There are several situations in which xml2po produces temporary
files.  If permissions in current directory are too restrictive, they
may cause mysterious breakages, though I have tried to make it act as
sanely as possible.

Normally, xml2po doesn't substitute external entities, because such
files should be translated by themselves.  Unfortunately, Python
bindings for libxml2 are not complete, and there's no simple way to
detect if entity is external or not.  Still, there's a method
debugDumpNode which can output relevant data to a filehandle (it
doesn't work with StringIO), so I decided to create a temporary file
.xml2po-checkingentities in the current working directory.  If it is
not writeable, instead of showing an error, xml2po will replace every
entity.

Option "-p file.po" depends on program msgfmt being available and in
the path.  If that fails, you'll see an error.



ChangeLog
2009-04-02  Claude Paroz  <claude@2xlibre.net>

	* modes/docbook.py: Use new hashlib module rather than deprecated md5
	module, when available.
	Fixes bug #576161.

2008-07-14  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py: Apply Goran Rakić's patch for #437371.

2007-03-16  Loïc Minier  <lminier@svn.gnome.org>

	* Makefile.am:
	Display xsltproc invocation, honor its exit status, and pull the
	source from srcdir; this might expose some bugs; bug #169324.

2007-03-02  Danilo Šegan  <danilo@gnome.org>

	* Fix #413602: Use mfile variable name instead of file, thus
	overwriting file() call used later.
	Patch by Eduardo Habkost <ehabkost@raisama.net>.

2006-12-26  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py (autoNodeIsFinal): Cache autoNodeIsFinal value.
	(worthOutputting): Cache worthOutputting value.

2006-12-26  Danilo Šegan  <danilo@gnome.org>

	Fix bug #343749.
	Support for XHTML mode, partially by Claude Paroz <paroz@email.ch>.

	* TODO: Updated.

	* xml2po.py (processAttribute): Added.
	(read_treatedattributes): Added.
	(processElementTag): Use treated attributes.

	* modes/Makefile.am (common_DATA): Added xhtml.py.

	* modes/gs.py (gsXmlMode.getTreatedAttributes):
	* modes/empty.py (emptyXmlMode.getTreatedAttributes):
	* modes/docbook.py (docbookXmlMode.getTreatedAttributes):
	Added new method to every other mode.

	* modes/xhtml.py: Added Claude's XHTML mode.

2006-12-26  Danilo Šegan  <danilo@gnome.org>

	Fix bug #378073.

	* xml2po.py (autoNodeIsFinal): Really ignore ignored tags.

2006-12-26  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py: Update copyright notice.

	* Makefile.am (xml2po): Remove configure.ac dependency.
	* autogen.sh, configure.ac: Removed.

2006-12-26  Danilo Šegan  <danilo@gnome.org>

	Fix bug #381607.

	* xml2po.py (MessageOutput.outputAll): Ignore untranslated
	messages in -r (reuse) mode.

2006-10-20  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py: Improve speed with Alexander Nedotsukov's patch from
	bug #363451.

2006-05-14  Danilo Šegan  <danilo@gnome.org>

	* modes/docbook.py (docbookXmlMode.postProcessXmlTranslation):
	Treat year as utf-8 as well.

2006-04-11  Danilo Šegan  <danilo@gnome.org>

        * modes/docbook.py: Check if "translators-credits" is translated.

2006-04-01  Danilo Šegan  <danilo@gnome.org>

	* tests/*: updated test cases.

2006-04-01  Danilo Šegan  <danilo@gnome.org>

	Added Ubuntu mode supporting &language; entity replacement.

	* modes/Makefile.am (common_DATA): Added ubuntu.py.
	* modes/ubuntu.py: Added.

2006-04-01  Danilo Šegan  <danilo@gnome.org>

	Fix #172459.

	* xml2po.py (NoneTranslations): return None for all *gettext calls
	as a fallback.
	(getTranslation): make use of NoneTranslations.
	(processElementTag): if mark_untranslated, node.setLang('C').

2006-04-01  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py (myAttributeSerialize): Added entity-non-expanding
	serialization.
	(startTagForNode): Use myAttributeSerialize.

2006-04-01  Danilo Šegan  <danilo@gnome.org>

	Fix #317110.

	* xml2po.py (startTagForNode): fix attribute handling.
	(replaceNodeContentsWithText): Use startTagForNode for starttag.

2006-04-01  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py (replaceNodeContentsWithText): Improve fix for #336765.

2006-03-31  Danilo Šegan  <danilo@gnome.org>

	Fixes bug #317502.

	* xml2po.py (normalizeNode): don't remove blank nodes followed by
	entities.

2006-03-31  Danilo Šegan  <danilo@gnome.org>

	Fixes bug #336765.

	* xml2po.py (replaceNodeContentsWithText): integrate DTD when
	keeping entities.
	* TODO: Document this fix.

2006-03-31  Danilo Šegan  <danilo@gnome.org>

	Fixes bug #172784.

	* xml2po.py (isFinalNode): ignore blank nodes.

2005-09-13  Stepan Kasal  <kasal@ucw.cz>

	* xml2po.pc.in: Add datarootdir variable, for compatibility
	with future Autoconf releases.

2005-07-23  Danilo Šegan  <danilo@gnome.org>

	* xml2po.1.xml: Document --language option, rewrite other parts.

	* xml2po.py: Add "--language=LANG" option to explicitely set language.

2005-07-23  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py: Replace /dev/null in msgfmt calls with NULL_STRING
	which is "NUL" when /dev/null doesn't exist (for Windows compatibility).

2005-07-22  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py: use mofile for getting translationlanguage in -t
	(reported by Daniel Leidert).

2005-07-20  Danilo Šegan  <danilo@gnome.org>

	* xml2po.py: remove path from language name.

2005-07-18  Danilo Šegan  <danilo@gnome.org>

	Fixes #310754.

	* modes/docbook.py (docbookXmlMode.postProcessXmlTranslation):
	Encode translator credits as UTF-8 for libxml2.

2005-06-26  Danilo Šegan  <danilo@gnome.org>

	Fixes #300442.

	* Makefile.am (xml2po.1): Use xsltproc to generate.
	* xml2po.1.xml: Add stylesheet declaration.

2005-04-02  Danilo Šegan  <dsegan@gmx.net>

	* xml2po.py (getTranslation): Use ugettext() instead of gettext().
	s/.serialize()/.serialize('utf-8')/g
	(replaceNodeContentsWithText): Use utf-8.

2005-04-01  Danilo Šegan  <dsegan@gmx.net>

	* xml2po.py (getTranslation): decode lookup string from utf-8.

2005-03-27  Danilo Šegan  <dsegan@gmx.net>

	* TODO (BUGS): Updated.

2005-03-27  Danilo Šegan  <dsegan@gmx.net>

	* xml2po.py (normalizeNode): Fix crash with "-k".
	(normalizeString): Add DTD always.

2005-03-27  Danilo Šegan  <dsegan@gmx.net>

	Complete fix of #162089.

	* xml2po.py (startTagForNode): Use .serialize() instead of doing
	it manually.

2005-03-27  Danilo Šegan  <dsegan@gmx.net>

	* xml2po.py (worthOutputting): Correct final tags semantics.
	(processElementTag): Use better check for final tags.

	* modes/docbook.py (docbookXmlMode.__init__): Separate objects tags.
	(docbookXmlMode.getIgnoredTags): Ignore them.
	(docbookXmlMode.getFinalTags): Use self.objects.

2005-03-27  Danilo Šegan  <dsegan@gmx.net>

	Fix #162089.

	* xml2po.py (replaceNodeContentsWithText): Don't use
	startTagForNode, use endTagForNode.

2005-03-27  Danilo Šegan  <dsegan@gmx.net>


tests/test.py
#!/usr/bin/env python
import sys, os, time

SIMPLETESTS = { 'deep-finals.xml' : {},
                'deep-nonfinals.xml': {},
                'attribute-entities.xml': {},
                'docbook.xml' : {},
                'mallard.xml' : {'options': '-m mallard' },
                'utf8-original.xml': {},
                'footnotes.xml': {},
                'keepents.xml': { "options" : "-k" },
                'adjacent-ents.xml': { "options" : "-k" },
                'ubuntu-mode.xml': { "options" : "-m ubuntu -k -l sr" },
                'xhtml.xml': { "options" : "-m xhtml" },
                }

OTHERTESTS = [ ('relnotes', 'test.sh') ]

if len(sys.argv) > 1:
    input = sys.argv[1]
    pot = input.replace(".xml", ".pot")
    po = input.replace(".xml", ".po")
    myopts = ""
    if len(sys.argv) > 2:
        for opt in sys.argv[2:]:
            myopts += " " + opt
    output = input.replace(".xml", ".xml.out")
    fullcommand = "../xml2po/xml2po %s %s | sed 's/\"POT-Creation-Date: .*$/\"POT-Creation-Date: \\\\n\"/' | diff -u %s -" % (myopts, input, pot)
    #print >>sys.stderr, fullcommand
    ret = os.system(fullcommand)
    if ret:
        print "Problem: extraction from '%s'" % (input)
    fullcommand = "../xml2po/xml2po -p %s %s %s | diff -u %s -" % (po, myopts, input, output)
    #print >>sys.stderr, fullcommand
    ret = os.system(fullcommand)
    if ret:
        print "Problem: merging translation into '%s'" % (input)
else:
    time_start = time.time()
    for t in SIMPLETESTS:
        myopts = SIMPLETESTS[t].get("options", "")
        if os.system("%s %s %s" % (sys.argv[0], t, myopts)):
            print "WARNING: Test %s failed." % (t)

    for t in OTHERTESTS:
        if os.system("cd %s && ./%s" % (t[0], t[1])):
            print "WARNING: Test %s failed." % (t[0])
    time_end = time.time()
    print "%d tests executed in %2.2f secs" % (len(SIMPLETESTS)+len(OTHERTESTS), time_end-time_start)


src/xml2pong/modes/gs.py
# Copyright (c) 2004, 2005, 2006 Danilo Segan <danilo@gnome.org>.
#
# This file is part of xml2po.
#
# xml2po is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# xml2po is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with xml2po; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

# Special case Gnome Summary
#

from .basic import basicXmlMode

class gsXmlMode(basicXmlMode):
    """Abstract class for special handling of document types."""
    def getIgnoredTags(self):
        "Returns array of tags to be ignored."
        return [ 'link', 'bug-stats', 'cvs-stats', 'unixname' ]

    def getFinalTags(self):
        "Returns array of tags to be considered 'final'."
        return ['title', 'para', 'name', 'desc' ]

    def _find_salute(self, node):
        if node.name == 'salute':
            return node
        child = node.children
        while child:
            ret = self._find_salute(child)
            if ret:
                return ret
            child = child.next
        return None

    def postProcessXmlTranslation(self, doc, language, translators):
        """Sets a language and translators in "doc" tree.

        "translators" is a string consisted of translator credits.
        "language" is a simple string.
        "doc" is a libxml2.xmlDoc instance."""
        # Find "salute" tag and add a comment about translator.
        if translators == self.getStringForTranslators():
            return
        else:
            salute = self._find_salute(doc.getRootElement())
            if salute and salute.name=='salute':
                # found
                salute.newChild(None, "para", translators)
            else:
                return

    def getStringForTranslators(self):
        """Returns None or a string to be added to PO files.

        Common example is 'translator-credits'."""
        return "translator-credits"

    def getCommentForTranslators(self):
        """Returns a comment to be added next to string for crediting translators.

        It should explain the format of the string provided by getStringForTranslators()."""
        return """Translators: enter a short paragraph creditting yourself and others who helped you with the translation."""


src/xml2pong/modes/xhtml.py
# Copyright (c) 2004, 2005, 2006 Danilo Segan <danilo@gnome.org>.
# Copyright (c) 2006, 2009 Claude Paroz <claude@2xlibre.net>.
#
# This file is part of xml2po.
#
# xml2po is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# xml2po is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with xml2po; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

# This implements special instructions for handling XHTML documents
# in a better way, particularly to extract some attributes in HTML tags

from .basic import basicXmlMode

class xhtmlXmlMode(basicXmlMode):
    """Class for special handling of XHTML document types."""
    def getFinalTags(self):
        "Returns array of tags to be considered 'final'."
        return ['p', 'li', 'pre']

    def getTreatedAttributes(self):
        "Returns array of tag attributes which content is to be translated"
        return ['title','alt']

    def getSpacePreserveTags(self):
        "Returns array of tags in which spaces are to be preserved."
        return ['pre']


src/xml2pong/modes/ubuntu.py
# Copyright (c) 2006 Danilo Segan <danilo@kvota.net>.

import libxml2

from .docbook import docbookXmlMode

class ubuntuXmlMode (docbookXmlMode):
    """Special-casing Ubuntu DocBook website documentation."""
    def postProcessXmlTranslation(self, doc, language, translators):
        """Sets a language and translators in "doc" tree."""

        # Call ancestor method
        docbookXmlMode.postProcessXmlTranslation(self, doc, language, translators)

        try:
            child = doc.docEntity('language')
            newent = doc.addDocEntity('my_internal_hacky_language_entity_decl', libxml2.XML_INTERNAL_GENERAL_ENTITY, None, None, language)
            newent.setName('language')
            child.replaceNode(newent)
        except:
            newent = doc.addDocEntity('language', libxml2.XML_INTERNAL_GENERAL_ENTITY, None, None, language)


src/xml2pong/modes/mallard.py
# -*- coding: utf-8 -*-
# Copyright (c) 2004, 2005, 2006 Danilo Segan <danilo@gnome.org>.
# Copyright (c) 2009 Shaun McCance <shaunm@gnome.org>
#
# This file is part of xml2po.
#
# xml2po is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# xml2po is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with xml2po; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

# This implements special instructions for handling DocBook XML documents
# in a better way.
#
#  This means:
#   — better handling of nested complicated tags (i.e. definitions of
#     ignored-tags and final-tags)
#   — support for merging translator-credits back into DocBook articles
#   — support for setting a language
#

import re
import libxml2
import os
import sys
try:
    # Hashlib is new in Python 2.5
    from hashlib import md5 as md5_new
except ImportError:
    from md5 import new as md5_new

from .basic import basicXmlMode

class mallardXmlMode(basicXmlMode):
    """Class for special handling of Mallard document types."""
    def __init__(self):
        pass

    def isBlockContext(self, node):
        if node.type == 'element' and node.parent.type == 'element':
            if node.parent.name in ('section', 'example', 'comment', 'figure',
                                    'listing', 'note', 'quote', 'synopsis'):
                return True
            elif node.parent.name == 'media':
                return self.isBlockContext(node.parent)
        return False

    def isFinalNode(self, node):
        if node.type == 'element':
            # Always block
            if node.name in ('p', 'screen', 'title', 'desc', 'cite', 'item'):
                return True
            # Always inline
            elif node.name in ('app', 'cmd', 'em', 'file', 'gui', 'guiseq', 'input',
                               'key', 'keyseq', 'output', 'span', 'sys', 'var'):
                return False
            # Block or inline
            elif node.name in ('code', 'media'):
                return self.isBlockContext(node)
            # Inline or info
            elif node.name == 'link':
                return node.parent.name == 'info'
        return False

    def getIgnoredTags(self):
        "Returns array of tags to be ignored."
        return []

    def getFinalTags(self):
        "Returns array of tags to be considered 'final'."
        return []

    def getSpacePreserveTags(self):
        "Returns array of tags in which spaces are to be preserved."
        return ['code', 'screen']

    def getStringForTranslators(self):
        """Returns string which will be used to credit translators."""
        return "translator-credits"

    def getCommentForTranslators(self):
        """Returns a comment to be added next to string for crediting translators."""
        return """Put one translator per line, in the form of NAME <EMAIL>, YEAR1, YEAR2"""

    def _md5_for_file(self, filename):
        hash = md5_new()
        input = open(filename, "rb")
        read = input.read(4096)
        while read:
            hash.update(read)
            read = input.read(4096)
        input.close()
        return hash.hexdigest()

    def _output_images(self, node, msg):
        if node and node.type=='element' and node.name=='media':
            attr = node.prop("src")
            if attr:
                dir = os.path.dirname(msg.filename)
                fullpath = os.path.join(dir, attr)
                if os.path.exists(fullpath):
                    hash = self._md5_for_file(fullpath)
                else:
                    hash = "THIS FILE DOESN'T EXIST"
                    print("Warning: image file '%s' not found." % fullpath, file=sys.stderr)

                msg.outputMessage("@@image: '%s'; md5=%s" % (attr, hash), node.lineNo(),
                                  "When image changes, this message will be marked fuzzy or untranslated for you.\n"+
                                  "It doesn't matter what you translate it to: it's not used at all.")
        if node and node.children:
            child = node.children
            while child:
                self._output_images(child,msg)
                child = child.next


    def preProcessXml(self, doc, msg):
        """Add additional messages of interest here."""
        root = doc.getRootElement()
        self._output_images(root,msg)

    def postProcessXmlTranslation(self, doc, language, translators):
        """Sets a language and translators in "doc" tree."""
        root = doc.getRootElement()
        if root and root.name == 'page':
            root.setProp('xml:lang', language)


src/xml2pong/modes/docbook.py
# -*- coding: utf-8 -*-
# Copyright (c) 2004, 2005, 2006 Danilo Segan <danilo@gnome.org>.
#
# This file is part of xml2po.
#
# xml2po is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# xml2po is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with xml2po; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

# This implements special instructions for handling DocBook XML documents
# in a better way.
#
#  This means:
#   — better handling of nested complicated tags (i.e. definitions of
#     ignored-tags and final-tags)
#   — support for merging translator-credits back into DocBook articles
#   — support for setting a language
#

# We use "currentXmlMode" class name for all modes
#  -- it might be better to have it named docbookXmlMode, but it will make loading harder;
#     it is also not necessary until we start supporting extracting strings from more
#     than one document type at the same time
#
import re
import libxml2
import os
import sys
from hashlib import md5 as md5_new
from .basic import basicXmlMode

class docbookXmlMode(basicXmlMode):
    """Class for special handling of DocBook document types.

    It sets lang attribute on article elements, and adds translators
    to articleinfo/copyright."""
    def __init__(self):
        self.lists = ['itemizedlist', 'orderedlist', 'variablelist',
                      'segmentedlist', 'simplelist', 'calloutlist', 'varlistentry' ]
        self.objects = [ 'figure', 'textobject', 'imageobject', 'mediaobject',
                         'screenshot' ]

    def getIgnoredTags(self):
        "Returns array of tags to be ignored."
        return  self.objects + self.lists

    def getFinalTags(self):
        "Returns array of tags to be considered 'final'."
        return ['para', 'formalpara', 'simpara',
                'releaseinfo', 'revnumber', 'title',
                'date', 'term', 'programlisting'] + self.objects + self.lists

    def getSpacePreserveTags(self):
        "Returns array of tags in which spaces are to be preserved."
        return [
            'classsynopsisinfo',
            'computeroutput',
            'funcsynopsisinfo',
            'literallayout',
            'programlisting',
            'screen',
            'synopsis',
            'userinput'
            ]

    def getStringForTranslators(self):
        """Returns string which will be used to credit translators."""
        return "translator-credits"

    def getCommentForTranslators(self):
        """Returns a comment to be added next to string for crediting translators."""
        return """Put one translator per line, in the form of NAME <EMAIL>, YEAR1, YEAR2"""

    def _find_articleinfo(self, node):
        if node.name == 'articleinfo' or node.name == 'bookinfo':
            return node
        child = node.children
        while child:
            ret = self._find_articleinfo(child)
            if ret:
                return ret
            child = child.next
        return None

    def _find_lastcopyright(self, node):
        if not node.children:
            return None
        last = node.lastChild()
        tmp = last
        while tmp:
            if tmp.name == "copyright":
                last = tmp
                break
            tmp = tmp.prev
        return last

    def _md5_for_file(self, filename):
        hash = md5_new()
        input = open(filename, "rb")
        read = input.read(4096)
        while read:
            hash.update(read)
            read = input.read(4096)
        input.close()
        return hash.hexdigest()

    def _output_images(self, node, msg):
        if node and node.type=='element' and node.name=='imagedata':
            # Use .fileref to construct new message
            attr = node.prop("fileref")
            if attr:
                dir = os.path.dirname(msg.filename)
                fullpath = os.path.join(dir, attr)
                if os.path.exists(fullpath):
                    hash = self._md5_for_file(fullpath)
                else:
                    hash = "THIS FILE DOESN'T EXIST"
                    print("Warning: image file '%s' not found." % fullpath, file=sys.stderr)

                msg.outputMessage("@@image: '%s'; md5=%s" % (attr, hash), node.lineNo(),
                                  "When image changes, this message will be marked fuzzy or untranslated for you.\n"+
                                  "It doesn't matter what you translate it to: it's not used at all.")
        elif node and node.children:
            child = node.children
            while child:
                self._output_images(child,msg)
                child = child.next


    def preProcessXml(self, doc, msg):
        """Add additional messages of interest here."""
        root = doc.getRootElement()
        self._output_images(root,msg)

    def postProcessXmlTranslation(self, doc, language, translators):
        """Sets a language and translators in "doc" tree.

        "translators" is a string consisted of "Name <email>, years" pairs
        of each translator, separated by newlines."""

        root = doc.getRootElement()
        # DocBook documents can be something other than article, handle that as well in the future
        while root and root.name != 'article' and root.name != 'book':
            root = root.next
        if root and (root.name == 'article' or root.name == 'book'):
            root.setProp('lang', language)
        else:
            return

        if translators == self.getStringForTranslators():
            return
        elif translators:
            # Now, lets find 'articleinfo' (it can be something else, but this goes along with 'article')
            ai = self._find_articleinfo(root)
            if not ai:
                return

            # Now, lets do one translator at a time
            lines = translators.split("\n")
            for line in lines:
                line = line.strip()
                match = re.match(r"^([^<,]+)\s*(?:<([^>,]+)>)?,\s*(.*)$", line)
                if match:
                    last = self._find_lastcopyright(ai)
                    copy = libxml2.newNode("copyright")
                    if last:
                        copy = last.addNextSibling(copy)
                    else:
                        ai.addChild(copy)
                    if match.group(3):
                        copy.newChild(None, "year", match.group(3).encode('utf-8'))
                    if match.group(1) and match.group(2):
                        holder = match.group(1)+"(%s)" % match.group(2)
                    elif match.group(1):
                        holder = match.group(1)
                    elif match.group(2):
                        holder = match.group(2)
                    else:
                        holder = "???"
                    copy.newChild(None, "holder", holder.encode('utf-8'))

# Perform some tests when ran standalone
if __name__ == '__main__':
    test = docbookXmlMode()
    print("Ignored tags       : " + repr(test.getIgnoredTags()))
    print("Final tags         : " + repr(test.getFinalTags()))
    print("Space-preserve tags: " + repr(test.getSpacePreserveTags()))

    print("Credits from string: '%s'" % test.getStringForTranslators())
    print("Explanation for credits:\n\t'%s'" % test.getCommentForTranslators())
